### PR TITLE
Make client usable with an Arc and make it usable between threads

### PR DESCRIPTION
**Pull request by osobiehl**
This PR allows the client to send requests using the same transport. With this PR, you can now `clone` the client or wrap it around an  `Arc` 

However, this means that you need to set an appropriate token if you want to send multiple requests at the same time.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -38,127 +48,248 @@ use async_trait::async_trait;
...
+    pub async fn send_packet(&self, packet: &Packet) -> IoResult<Packet> {
+        self.do_request_response_for_packet(packet).await
     }
-    pub fn from_transport(transport: T) -> Self {
+
+    pub async fn do_request_response_for_packet(&self, packet: &Packet) -> IoResult<Packet> {
+        let mut receiver = self.establish_receiver_for(packet).await;
+        let result = self
+            .do_request_response_for_packet_inner(packet, &mut receiver)
+            // TODO: join into single request instead of racily re-requesting
```
Comment (Covertness) [inline src/client.rs:283]: Is it still waiting TODO ?

File: src/client.rs [hunk 2]
```diff
@@ -263,7 +405,27 @@ impl UdpCoAPClient {
...
     pub fn set_broadcast(&self, value: bool) -> IoResult<()> {
-        self.transport.inner.socket.set_broadcast(value)
+        self.transport.transport.socket.set_broadcast(value)
+    }
+
+    /// creates a receiver based on a specific request
+    /// this method can be used if you send a multicast request and
+    /// expect multiple responses.
+    /// only use this method if you know what you are doing
+    pub async fn create_receiver_for(
```
Comment (Covertness) [inline src/client.rs:415]: I think it would be better if you can provide an example how to use the function to handle multiple responses

**Review by Covertness — CHANGES_REQUESTED** [review]

File: src/client.rs (hunk 1 above)
Comment (osobiehl) [reply src/client.rs:283]: it was outdated :) will remove

**Comment by osobiehl** [comment]
I've gone ahead and made more changes so that you can now fully use the client in an `Arc`

File: src/client.rs (hunk 2 above)
Comment (osobiehl) [reply src/client.rs]: added example!

**Pull request merged**